Added overload to support SDK supplying query string on invoked URL

WhitWaldo opened the pull request:
# Description
When a method is invoked via the Dapr service invocation building block, a query string can be passed through on the invocation URL without issue. However, when using the .NET SDK, this behavior isn't exposed anywhere. Existing overloads allow the app ID to be specified, the method named, the payload to be provided, but that's it - it's up to the developer to realize that a query string is supported, then augment the `RequestUri` property of the returned `HttpRequestMessage` to append their query string values, and proceed with the invocation.

This PR provides overloads to the invocation methods to allow an `IReadOnlyCollection<KeyValuePair<string,string>>` to be provided as the collection of key/value pairs to build/supplement the query string with. This uses a static extension method I've contributed (and added tests for) that appends the encoded key/value pairs provided in the collection to build a standards-matching query string so developers don't have to maintain the same on their own.

Why an `IReadOnlyCollection` instead of a `Dictionary<string,string>`? Because identical keys [can exist more than once in a query string](https://url.spec.whatwg.org/#example-searchparams-sort) with different values, so this needs to be more of a collection and less of a mapping. And because at the end of the day, this is being written to a string-based URI, these methods expect string-based keys and values, leaving it to the developer to figure out how to serialize accordingly before calling this.

Finally, I had the existing concrete implementations point to the new overloads and pass empty collections so as to avoid code duplication.

## Issue reference
Please reference the issue this PR will close: #1303 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

Documentation PR extended here, though generically and not in a C#-specific way: [https://github.com/dapr/docs/pull/4217](https://github.com/dapr/docs/pull/4217)

philliphoff commented on src/Dapr.Client/Extensions/HttpExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        /// <param name="message">The message to append the query string parameters to.</param>
+        /// <param name="queryStringParameters">The key/value pairs to populate the query string with.</param>
+        public static void AddQueryParameters(this HttpRequestMessage? message,
```
A couple of comments:

 - I feel like `HttpRequestMessage` is too high-level for this; IMO it'd be better to accept/return a `Uri`, which also avoids mutation of the message, which I find a little odd for a method that's not otherwise part of a "builder" pattern.
 - I'd make this class/method `internal` if not made part of `DaprClientGrpc` itself, lest developers start using it as a general-purpose query management utility function and commit us to support the API into the future.

WhitWaldo replied:
Fair on both counts. Updated the extension, the method and the unit tests.

philliphoff reviewed (changes requested):
@WhitWaldo  It looks like the integration test failures are due to something in this PR (as they succeed in `master`). Can you take a look and see what's going on?

WhitWaldo commented:
@philliphoff I've updated my branch and resynced my local branch to master. I'm running the tests locally and everything I contributed here is testing fine. I'm showing errors when testing locally that appear entirely unrelated to my contribution:

- InvokeMethodAsync_VoidResponseNoHttpMethod_Success, InvokeMethodAsync_VoidresponseWithHttpMethod_Success, InvokeMethodAsync_VoidVoidNotHttpMethod_Success and InvokeMethodAsync_VoidVoidWithHttpMethod_Success are all failing because of "The client has 1 or more incomplete requests. Use `request.Dismiss()` if the test is uninterested in the response." which suggests some sort of shared client issue?
- Seeing an error with TestWorkflowLogging that's failing because the file is in use by another process

I don't necessarily mind tackling either of these, but I handling them in the context of this PR may be out of scope. I'll keep poking at it and see what my changes might have to do with the error.

philliphoff commented:
@WhitWaldo I agree that I don't see an immediate connection to your changes, but since the tests only seem to fail in this branch, we can't commit it until that's sorted out.

frankbuckley commented on src/Dapr.Client/Extensions/HttpExtensions.cs:
```diff
@@ -0,0 +1,51 @@
+        /// <param name="uri">The uri to append the query string parameters to.</param>
+        /// <param name="queryStringParameters">The key/value pairs to populate the query string with.</param>
+        public static Uri AddQueryParameters(this Uri? uri,
+            IReadOnlyCollection<KeyValuePair<string, string>>? queryStringParameters)
```
Curious why not [NameValueCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.namevaluecollection?view=net-8.0#remarks)?

WhitWaldo replied:
Fundamentally, I was looking to ensure that a read-only collection is passed. Conceptually, they wind up operationally working the same way. I don't see why an overload couldn't be added for a `NameValueCollection` as well.

WhitWaldo commented:
@philliphoff I show that all the tests are passing as of the last commit. Figured out the issue - there was some ambiguity about which method should be invoked given all the overloads available. One of them, added as part of this issue, added:
```cs
public override HttpRequestMessage CreateInvokeMethodRequest(HttpMethod httpMethod, string appId, string methodName,
    IReadOnlyCollection<KeyValuePair<string, string>> queryStringParameters)
```

The existing overload was:
```cs
public override HttpRequestMessage CreateInvokeMethodRequest<TRequest>(HttpMethod httpMethod, string appId, string methodName, TRequest data)
```

And at runtime then, it was taking the query string parameters and instead dropping them in as the request payload. I've remedied by changing the latter overload signatures to also accept the `IReadOnlyCollection<KeyValuePair<string, string>>` and simply pass it an empty collection when it's not provided and both locally and in the CI/CD, everything looks to be working now.
